crates/runc-shim/src/cgroup_memory.rs

```diff
@@ -0,0 +1,126 @@
+use containerd_shim::{
+    error::{Error, Result},
+    io_error, other_error,
+};
+use std::path::Path;
+use tokio::{
+    fs::{self, File},
+    io::{AsyncBufReadExt, AsyncReadExt, BufReader},
+    sync::mpsc::{self, Receiver},
+};
+
+use nix::sys::eventfd;
+use std::os::unix::io::{AsRawFd, FromRawFd};
+
+pub async fn get_path_from_cgorup(pid: u32) -> Result<String> {
+    let proc_path = format!("/proc/{}/cgroup", pid);
+    let file = File::open(proc_path.clone())
+        .await
+        .map_err(io_error!(e, "open {}.", proc_path))?;
+    let mut lines = BufReader::new(file).lines();
```

Comment:
Let's make this code a bit more Rusty:

```rust
    let proc_path = format!("/proc/{}/cgroup", pid);
    let (_, path) = read_to_string(&proc_path)
        .await
        .map_err(io_error!(e, "open {}.", proc_path))?
        .lines()
        .find(|line| line.contains("memory")) // Only one line contains "memory"
        .ok_or(Error::Other("Memory line not found".into()))?
        .split_once(":memory:") // Split by ":memory:"
        .ok_or(Error::Other("Failed to parse memory line".into()))?;

    Ok(path)
```

(I haven't tested it)

Reply:
Alright, I’ll give it a try and make the necessary modifications.